docs/index.md

```diff
@@ -1,23 +1,164 @@
+- syncing data, for example [using a sync engine like ElectricSQL](#)
+- storing local data, for example [in-memory client data or UI state](#)
+- from live queries, creating [derived collections as materialised views](#)
+
+Once you have your data in collections, you can query across them to bind results to your components. This allows you to:
+
+1. load normalised data into collections and then de-normalise it through queries; simplifying your backend by avoiding the need for bespoke API endpoints that match your client
+2. join data from multiple sources; for example, syncing some data out of a database, fetching some other data from an external API and then joining these into a unified data model for your front-end code
+
+Your components don't need to know where the data in the collections they're working with came from.
+
+### Using live queries
+
+Live queries are used to query data out of collections and bind the results to state variables in your components. Live queries are reactive: when the underlying data changes in a way that would affect the query result, the new result is automatically assigned to the state variable, triggering a re-render.
+
+TanStack DB live queries are implemented using [d2ts](https://github.com/electric-sql/d2ts), a Typescript implementation of differential dataflow. This allows the query results to update *incrementally* (rather than by re-running the whole query). This makes them blazing fast, usually sub-millisecond, even for highly complex queries.
+
+Live queries support joins across collections. In fact, live queries return new collections. It's collections all the way down.
+
+### Making local writes
+
+Collections support `insert`, `update` and `delete` operations. These operations must be made within the context of a transactional mutator. Rather than mutating the collection data directly, the collection internally treats its synced/loaded data as immutable and maintains a seperate set of local mutations as optimistic state.
+
+When live queries read from the collection, they see a local view that overlays the local optimistic mutations on-top-of the immutable synced data.
+
+In addition, the local mutations are passed to the async `mutationFn` that's passed in when creating the mutator.
+
+```ts
+const addTodo = useOptimisticMutation({ mutationFn })
+```
+
+This mutationFn is responsible for handling the writes, usually by sending them to a server or a database. TanStack DB waits for the function to resolve before then removing the optimistic state that was applied when the local writes was made.
+
+For example, in the following code, the mutationFn first sends the write to the server using `await api.todos.create(newTodo)` and then calls `await collection.invalidate()` to trigger a re-fetch of the collection contents using TanStack Query. When this second await resolves, the collection is up-to-date with the latest changes and the optimistic state is safely discarded.
```

Comment:
I'd add something about how the collection blocks applying updates while this function runs so that the optimistic state is dropped at the exact same time as the backend state is applied so there's never any flicker in the UI.